Rewrite Local()

vladimir-ch opened the pull request:
The first step towards fixing `Local()`, `minimize()` and `Linesearch`. I tried to keep the changes limited to `Local()`, but it was not possible everywhere.

Summary of changes:

* Put various initializations into a more reasonable order, so that things are initialized only when necessary.
* Properly record the starting location.
* Check the starting location for convergence before calling `minimize()`
* Check the initial function status if it is a `Statuser`
* Reduce the number of arguments to `minimize()`
* `getStartingLocation()` now also returns `EvaluationType` that was used at `initX`
* Add implementation comments to `Local()`

vladimir-ch commented:
Could you PTAL, @btracey ?

btracey commented on local.go:
```diff
@@ -54,59 +54,92 @@ func Local(f Function, initX []float64, settings *Settings, method Method) (*Res
-		err = settings.Recorder.Init(funcInfo)
+		// Runtime is the only Stats field that needs to be updated here.
+		stats.Runtime = time.Since(startTime)
+		// TODO: Replace NoIteration with InitialIterate when it is added.
```
InitialIteration

vladimir-ch replied:
or InitIteration?

btracey commented on local.go:
```diff
@@ -54,59 +54,92 @@ func Local(f Function, initX []float64, settings *Settings, method Method) (*Res
 	if settings.Recorder != nil {
-		err = settings.Recorder.Init(funcInfo)
+		// Runtime is the only Stats field that needs to be updated here.
+		stats.Runtime = time.Since(startTime)
```
Runtime can be before the if, as it's called both in the if and out of it.

vladimir-ch replied:
Changed

btracey commented on local.go:
```diff
@@ -54,59 +54,92 @@ func Local(f Function, initX []float64, settings *Settings, method Method) (*Res
-	status, err := minimize(settings, location, method, funcInfo, stats, funcs, optLoc, startTime)
+	// Check if the starting location satisfies the convergence criteria.
+	stats.Runtime = time.Since(startTime)
+	// TODO: Replace MajorIteration with InitialIteration when it is added and
```
TODO should have your handle with it.

TODO(vladimir-ch):

btracey replied:
(This applies globally)

vladimir-ch replied:
OK

btracey commented on local.go:
```diff
@@ -54,59 +54,92 @@ func Local(f Function, initX []float64, settings *Settings, method Method) (*Res
+		// evaluate.
+		// TODO: Is this the best place for checking the function status?
+		if funcInfo.IsStatuser {
+			// TODO: should the function really return status?
```
I think so. It allows the function to signal completion for some reason. The only other way to communicate that is by falsely returning 0 for the gradient (or something). Two good reasons to have this:
1) The function can have an error for some reason. This is a way to put in that check without having to do it at every function evaluation where there is nothing the optimizer can do about it
2) In machine learning, one often wants to converge until the testing error is minimized (not the training error). The machine learning parameter function can return a status signifying early convergence even though a local minima has not been found.

vladimir-ch replied:
OK, removed the line.

btracey commented on local.go:
```diff
@@ -54,59 +54,92 @@ func Local(f Function, initX []float64, settings *Settings, method Method) (*Res
+			// TODO: should the function really return status?
+			status, err = funcs.status.Status()
+			if err != nil || status != NotTerminated {
+				// Function status is bad, but at least we have the starting
```
I think this may be a good case for a goto. Have line 119 be a "Cleanup:" label and jump to it in this f. The recorder still can get useful information here if err != nil && status != NotTerminated.

btracey replied:
Status doesn't have to be bad. "// Function status says optimization should be terminated"

vladimir-ch replied:
goto and label Cleanup added, comment updated.

btracey commented on local.go:
```diff
@@ -346,6 +382,7 @@ func update(location *Location, optLoc *Location, stats *Stats, iterType Iterati
 		stats.MajorIterations++
 	}
+	// TODO: Update optLoc only at MajorIterations?
 	if location.F < optLoc.F {
```
This should be <=

vladimir-ch replied:
Done.

btracey commented on local.go:
```diff
@@ -138,6 +171,7 @@ func minimize(settings *Settings, location *Location, method Method, funcInfo *F
 
 		// Compute the new function and update the statistics
 		evaluate(funcs, funcInfo, evalType, xNext, location, stats)
+		// TODO: Check for NaN.
```
I don't think NaN/Inf is necessarily a problem. Getting it on major iterations is bad, but the linesearchers should be smart enough to backtrack form that (and if they aren't we should make them as such). NaN and Inf are a problem at the starting location because gradients don't make any sense.

vladimir-ch replied:
OK, I will add an issue to make sure linesearchers handle NaN well.

vladimir-ch commented:
I addressed the comments.

I added one check at the beginning of Local(): if f is a Statuser and Status() returns an error, panic.

PTAL, @btracey

btracey commented on local.go:
```diff
@@ -54,59 +54,100 @@ func Local(f Function, initX []float64, settings *Settings, method Method) (*Res
+		if err != nil {
+			// Function status is bad, panic.
+			// TODO: Document this panic in the function comment.
+			panic(fmt.Sprintf("bad initial function status: %v", err))
```
This doesn't feel to me like it should panic. Since an error value is returned , it can presumably be handled, which means it should be returned.

btracey replied:
Honestly though, I'm okay with not having this call at all. Right now the flow is simple: 1) Call the function to generate the initial location if necessary. 2) Check the status before each step of the minimization, returning the Status and error. Status is checked before the first step of the minimization, so all this does is catch an error if the initial condition is not specified. If I were a function writer, I would have the returned function value be NaN or Inf if the function was actually in error, both of which will create an error if no initial condition is specified. If the initial condition is specified, we already check Status before the first step and do the reasonable thing.

If you do want this check here, it should behave the same as all the other checks (i.e. go to Cleanup).

btracey replied:
"return nil, err" like in the Recorder check is better.

vladimir-ch replied:
The motivation for this check was to make sure that the function has a good status before calling getStartingLocation(). The motivation for the panic was that if we return an error, the caller should be able to handle it, so if he is able to handle it after, he can also handle it before and should not call Local() with bad function status in the first place. I do not insist on this, I put in there as a possibility and to discuss it.

I will change it to "return nil, err".

vladimir-ch replied:
Changed.

btracey commented on local.go:
```diff
@@ -54,59 +54,100 @@ func Local(f Function, initX []float64, settings *Settings, method Method) (*Res
+		// evaluate.
+		// TODO(vladimir-ch): Is this the best place for checking the function
+		// status? Revise after minimize() is updated.
+		if funcInfo.IsStatuser {
```
Is this check actually necessary? It already gets checked before the next function evaluation in minimize. Removing it would make this much simpler and would also allow the removal of the Cleanup label (sorry for not seeing this before).

vladimir-ch replied:
No, it is not necessary. I wanted to revisit this after reordering the calls in minimize(), but it is already clear that funcInfo.status will be checked before calling evaluate(). Removed.

btracey commented on local.go:
```diff
@@ -54,59 +54,100 @@ func Local(f Function, initX []float64, settings *Settings, method Method) (*Res
+	// is added and checkConvergence() updated to handle it.
+	status = checkConvergence(optLoc, MajorIteration, stats, settings)
+	if status == NotTerminated {
+		// The starting location is not good enough, we need to start the
```
need to perform a minimization.

vladimir-ch replied:
Changed.

btracey commented on local.go:
```diff
@@ -346,7 +389,8 @@ func update(location *Location, optLoc *Location, stats *Stats, iterType Iterati
 		stats.MajorIterations++
 	}
-	if location.F < optLoc.F {
+	// TODO: Update optLoc only at MajorIterations?
```
I don't think we want to do this. For example, a linesearch may find a better function value but not have sufficient gradient decrease to meet Armijio. It wants to continue the linesearch, but the function evaluation limit is reached. We want to return the best point found in this case I would imagine.

vladimir-ch replied:
SGTM

btracey commented on local.go:
```diff
@@ -260,6 +281,7 @@ func checkConvergence(loc *Location, itertype IterationType, stats *Stats, setti
 	}
 
 	if settings.Runtime > 0 {
+		// TODO: It would be nice to update Runtime here.
```
This also needs a (vladimir-ch)

btracey commented:
LGTM with the last comment addressed.
